drivers/input/keyboard/brain-kbd-i2c.c

```diff
@@ -22,143 +22,145 @@
+	unsigned int normal_event_code;
+	unsigned int symbol_event_code;
 };
 
 struct bk_i2c_data {
 	struct i2c_client *cli;
 	struct input_dev *idev;
-	struct keymap_def *km;
-	struct keymap_def *km_symbol;
-	int kmlen;
-	int kmlen_symbol;
+	struct keymap_def keymaps[BK_KEYCODE_MAX];
 
-	bool symbol;
+	bool symbol_states[BK_KEYCODE_MAX];
+	bool symbol_mode;
 	u32 symbol_keycode;
 	bool closing;
-	bool symbol_flag[256];
 };
 
-static bool detect_key(struct bk_i2c_data *kbd, u8 keycode)
+static bool handle_switch(struct bk_i2c_data *kbd, u8 keycode)
 {
-	int i;
-	if (BK_IS_SWITCH(keycode)) {
-		bool sw_on = BK_SWITCH_ON(keycode);
-		unsigned int sw_code;
-		switch (BK_SW_CODE(keycode)) {
-		case BK_SW_LCD_TRANSFORMING_TO_TABLET:
-			if (!sw_on) {
-				kbd->closing = false;
-			}
-			return true;
-		case BK_SW_LCD_TRANSFORMING_TO_CLOSED:
-			if (sw_on) {
-				kbd->closing = true;
-			}
-			return true;
-
-		case BK_SW_LCD_FULLY_TRANSFORMED:
-			sw_code = (kbd->closing) ? SW_LID : SW_TABLET_MODE;
-			input_report_switch(kbd->idev, sw_code, sw_on);
-			return true;
-
-		case BK_SW_USB_VBUS:
-			input_report_switch(kbd->idev, SW_DOCK, sw_on);
-			return true;
-		default:
-			dev_dbg(&kbd->cli->dev, "Unknown switch event %0x02X",
-				keycode);
-		}
-		return false;
+	bool sw_on = BK_SWITCH_ON(keycode);
+	unsigned int sw_code;
+
+	switch (BK_SW_CODE(keycode)) {
+	case BK_SW_LCD_TRANSFORMING_TO_TABLET:
+		if (!sw_on)
+			kbd->closing = false;
+		return true;
+
+	case BK_SW_LCD_TRANSFORMING_TO_CLOSED:
+		if (sw_on)
+			kbd->closing = true;
+		return true;
+
+	case BK_SW_LCD_FULLY_TRANSFORMED:
+		sw_code = (kbd->closing) ? SW_LID : SW_TABLET_MODE;
+		input_report_switch(kbd->idev, sw_code, sw_on);
+		return true;
+
+	case BK_SW_USB_VBUS:
+		input_report_switch(kbd->idev, SW_DOCK, sw_on);
+		return true;
+
+	default:
+		dev_dbg(&kbd->cli->dev, "Unknown switch event %0x02X", keycode);
 	}
+	return false;
+}
 
-	if ((keycode & 0x3f) == (u8)(kbd->symbol_keycode)) {
-		if (BK_IS_PRESSED(keycode)) {
-			dev_dbg(&kbd->cli->dev, "symbol pressed!\n");
-			kbd->symbol = true;
-		} else {
-			dev_dbg(&kbd->cli->dev, "symbol released!\n");
-			kbd->symbol = false;
+static bool handle_symbol_key(struct bk_i2c_data *kbd, u8 keycode)
+{
+	struct keymap_def *keymap = &kbd->keymaps[BK_KEY(keycode)];
+
+	if (keymap->symbol_event_code != KEY_RESERVED) {
+		if (kbd->symbol_states[BK_KEY(keycode)] == false) {
+			input_report_key(kbd->idev, keymap->normal_event_code, 0);
+
+			dev_dbg(&kbd->cli->dev,
+				"mode changed, normal key %02x(%02x) released\n",
+				BK_KEY(keycode), keymap->normal_event_code);
 		}
+
+		input_report_key(kbd->idev,
+			keymap->symbol_event_code, BK_IS_PRESSED(keycode));
+
+		kbd->symbol_states[BK_KEY(keycode)]  = BK_IS_PRESSED(keycode);
+
+		dev_dbg(&kbd->cli->dev, "symbol key %02x(%02x) %s\n",
+			BK_KEY(keycode), keymap->symbol_event_code,
+			BK_IS_PRESSED(keycode) ? "pressed": "released");
+
 		return true;
+	} else {
+		return false;
 	}
```

Comment:
Negating the condition makes the if block shorter.

```suggestion
	if (keymap->symbol_event_code == KEY_RESERVED) {
		return false;
	}

	if (kbd->symbol_states[BK_KEY(keycode)] == false) {
		input_report_key(kbd->idev, keymap->normal_event_code, 0);

		dev_dbg(&kbd->cli->dev,
			"mode changed, normal key %02x(%02x) released\n",
			BK_KEY(keycode), keymap->normal_event_code);
	}

	input_report_key(kbd->idev,
		keymap->symbol_event_code, BK_IS_PRESSED(keycode));

	kbd->symbol_states[BK_KEY(keycode)]  = BK_IS_PRESSED(keycode);

	dev_dbg(&kbd->cli->dev, "symbol key %02x(%02x) %s\n",
		BK_KEY(keycode), keymap->symbol_event_code,
		BK_IS_PRESSED(keycode) ? "pressed": "released");

	return true;
```